url49526af437_Maria_G_CS325_tgz-pJ8-assignments_farmer_FarmerImplJ8/src/minidraw/boardgame/BoardDrawing.java

```diff
@@ -0,0 +1,241 @@
+   expressed or implied. You may study, use, modify, and 
+   distribute it for non-commercial purposes. For any 
+   commercial use, see http://www.baerbak.com/
+ */
+public class BoardDrawing<LOCATION> extends StandardDrawing 
+  implements BoardGameObserver<LOCATION> {
+
+  /** "Map of list" collection, mapping each location to the set of
+   * images positioned on it.
+   */
+  protected Map<LOCATION, List<BoardFigure>> figureMap = null;
+
+  /** Map collection, mapping graphical (x,y) positions to
+   * the props of the game.
+   */
+  protected Map<String, BoardFigure> propMap = null;
+  
+  protected FigureFactory<LOCATION> factory;
+  protected PositioningStrategy<LOCATION> adjuster;
+  protected PropAppearanceStrategy propChanger;
+
+  /** Construct a Drawing specifically for handling figures
+   * associated with Board games.
+   * @param factory abstract factory to return the initial set of 
+   * piece images and props
+   * @param adjuster the strategy for calculating proper positions
+   * of piece images on locations. Must be non-null.
+   * @param propChanger the strategy for changing the appearance
+   * of props. This reference may be null if no props are used and
+   * the domain code ensures that no propChangeEvents are ever
+   * sent.
+   */
+  public BoardDrawing(FigureFactory<LOCATION> factory, 
+                      PositioningStrategy<LOCATION> adjuster,
+                      PropAppearanceStrategy propChanger ) {
+    super();
+    this.factory = factory;
+    this.adjuster = adjuster;
+    this.propChanger = propChanger;
+    buildPieceMap();
+    buildPropMap();
+  }
+  
+  protected void adjustFigurePosition(BoardFigure figure, 
+                                      LOCATION location,
+                                      int index) {
+    int cx, cy; // current (x,y)
+    Rectangle rect = figure.displayBox();
+    cx = rect.x; cy = rect.y;
+    Point properPosition;
+
+    properPosition = 
+      adjuster.calculateFigureCoordinatesIndexedForLocation(location, index);
+
+    int rx, ry;
+    rx = properPosition.x; ry = properPosition.y;
+    figure.moveBy( rx-cx, ry-cy);
+  }
+
+  protected void buildPieceMap() {
+    figureMap = factory.generatePieceMultiMap();
+    if ( figureMap == null ) {
+      throw new RuntimeException("BoardGame contract violation: buildPieceMap assumes"+
+          " a non-null map is return from the FigureFactory." );
+    }
+
+    for ( Entry<LOCATION,List<BoardFigure>> e : figureMap.entrySet() ) {
+      int index = 0;
+      for ( BoardFigure f : e.getValue()) {
+        adjustFigurePosition(f, e.getKey(), index++);
+        super.add(f);
+      }
+    }
+  }
+  protected void buildPropMap() {
+    propMap = factory.generatePropMap();
+    if ( propMap == null ) {
+      // create a null object to avoid if's all over the place.
+      propMap = new HashMap<String,BoardFigure>();
+    }
+    for ( String key : propMap.keySet()) {
+      Figure f = propMap.get(key);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.